### PR TITLE
fix: run precommit hook only on staged files

### LINKED ISSUE #376: Only run precommit hook on staged files
Sometimes, I work with unencrypted values locally because I am tweaking my `.env` files and do not intend to commit them.
Still, during commit, the hook will also verify `.env` files that are not staged, and that isn't very pleasant to encrypt/decrypt files to please the hook.  

We could check the diff in [this phase](https://github.com/dotenvx/dotenvx/blob/7a2b8f22045aef0335fe6cf53744cf4e40d3e0fb/src/lib/services/precommit.js#L42-L66) of the precommit (when `git` is available in user environment). What do you think?

### PR DESCRIPTION
**Pull request by robertsLando**
Fixes #376

### CODE REVIEW

**Comment by robertsLando** [comment]
The diff is a bit misleading, I just added an if inside the files foreach that checks first if the given file is going to be commited or not, just that. Checked and seems working

**Comment by robertsLando** [comment]
cc @getlarge @motdotla

File: src/lib/services/precommit.js [hunk 1]
```diff
@@ -77,9 +81,22 @@ class Precommit {
...
+      return false;
+    }
+  }
+
+  _installPrecommitHook() {
     return new InstallPrecommitHook().run()
   }
 }
 
+new Precommit().run()
```
Comment (getlarge) [inline src/lib/services/precommit.js:100]: oops

**Comment by robertsLando** [comment]
@motdotla A little off-topic note here. I see you are using pkg to bundle dotenvx, if you are interested I was a maintainer of pkg and after Vercel archived the repo I started my own fork that's actually the most maintained one with also all latest nodejs versions supported. You can find it here: https://github.com/yao-pkg/pkg 😄

File: src/lib/services/precommit.js [hunk 2]
```diff
@@ -77,9 +81,22 @@ class Precommit {
...
+    try {
+      const output = execSync('git diff --cached --name-only').toString()
+      const files = output.split('\n')
+      return files.includes(filePath)
+    } catch (error) {
+      console.error('Error checking if file is to be committed:', error)
+      return false
+    }
+  }
+
```
Comment (getlarge) [inline src/lib/services/precommit.js:94]: I'm not sure if it should be taken for granted that the user is using git...
Maybe the Precommit class should be configured to support several version control systems and default to git?

File: src/lib/services/precommit.js (hunk 1 above)
Comment (robertsLando) [reply src/lib/services/precommit.js:100]: image

File: src/lib/services/precommit.js (hunk 2 above)
Comment (robertsLando) [reply src/lib/services/precommit.js:94]: I agree but I would need suggestions so about how this is supposed to be fixed :( I could suggest to check if git is used and if not fallback to previous check

Comment (robertsLando) [reply src/lib/services/precommit.js]: I added a possible fix for this so at worse this will keep working like it is now

**Comment by motdotla** [comment]
>  I see you are using pkg to bundle dotenvx, if you are interested I was a maintainer of pkg and after Vercel archived the repo I started my own fork that's actually the most maintained one with also all latest nodejs versions supported. You can find it here: https://github.com/yao-pkg/pkg 😄

wait, what!? sick! 🤙

i'll switch over dotenvx-ext-vault to it first. if all goes well, we'll switch dotenvx to use it as well. thank you for taking that over. i really like pkg.

we're using this patch here: https://github.com/dotenvx/dotenvx/blob/main/patches/pkg%2B5.8.1.dev.patch

does yao-pkg solve for that currently?

File: src/lib/services/precommit.js (hunk 2 above)
Comment (motdotla) [reply src/lib/services/precommit.js]: that fallback is good. safe to safe that prob 90% of those using it currently are using git. likely even higher.

**Comment by robertsLando** [comment]
@motdotla Is that an issue related only to dotenvx or a general one? If the second I can include it without any problem 🙏🏻 Thanks for the support as well ♥️

**Comment by motdotla** [comment]
ok, i'll use 18 or 20. 

as far as I could tell it affects any project that might invoke itself

* the problem: https://github.com/dotenvx/dotenvx/issues/276#issuecomment-2198323533
* my fix for dotenvx: https://github.com/dotenvx/dotenvx/pull/279
* the core solution: https://github.com/vercel/pkg/issues/897#issuecomment-1663052209

**Comment by robertsLando** [comment]
Will take a look at that on Monday and let you know! 🙏🏻

**Comment by motdotla** [comment]
this PR for the precommit is released as `v1.14.1`

**Comment by robertsLando** [comment]
Thanks 🙏🏼

**Comment by robertsLando** [comment]
@motdotla Could also this fix work? https://github.com/yao-pkg/pkg/issues/70#issuecomment-2351358704

**Pull request merged**